Improve & update release process to reflect recent changes (#3242)
psf/black: docs/contributing/release_process.md

# Release process

_Black_ has had a lot of work done into standardizing and automating its release
process. This document sets out to explain what everything works and how to release
_Black_ using said automation.

## Release cadence

> **We aim to release whatever is on `main` every 1-2 months.** This ensures merged

Author: 1-2 months felt like the most reasonable given the current pace of Black's development, but I'm open to alternatives.

improvements and bugfixes are shipped to users reasonably quickly, while not massively
fracturing the user-base with too many versions. This also keeps the workload on
maintainers consistent and predictable.

If there's not much new on `main` to justify a release, it's acceptable to skip a

> month's release. Although, ideally January releases should not be skipped because as per

Author: Not entirely sure if we want to formalise this, but I feel like sometimes our development pace gets slow enough where releasing monthly stops being valuable.

our [stability policy](labels/stability-policy), the first release in a new calendar
year may make changes to the _stable_ style. While the policy applies to the first
release (instead of only January releases), confining changes to the stable style to
January will keep things predictable (and nicer) for users.

Unless there is a serious regression or bug that requires immediate patching, **there
should not be more than one release per month**. While version numbers are cheap,
releases require a maintainer to both commit to do the actual cutting of a release, but
also to be able to deal with the potential fallout post-release. Releasing more
frequently than monthly nets rapidly diminishing returns.

## Cutting a release

**You must have `write` permissions for the _Black_ repository to cut a release.**

The 10,000 foot view of the release process is that you publish a [GitHub Release] which
then triggers release automation that builds all release artifacts and publishes them to
the various platforms we publish to.

To cut a release:

> 1. Determine the release's version number

Author: I tried to make this as easy as possible to follow while being thorough, but I might have made this too explicit and made the release process look scarier than it really is. I'd appreciate thoughts on it

>    - **_Black_ follows the [CalVar] versioning standard using the `YY.M.N` format**

Reviewer: ```suggestion
   - **_Black_ follows the [CalVer] versioning standard using the `YY.M.N` format**
```

Author: Ah good catch!

   1. You can leave the release title blank, GitHub will default to the tag name
   1. Paste the changelog Markdown you copied earlier into the description box
1. Publish the GitHub Release you just drafted
   - On publication, the release automation will trigger and run on the new tag
1. At this point, you're basically done. It's good practice to go and watch to make sure
   all the [release workflows][black-actions] pass, although you will receive a GitHub
   notification if something does fail.
   - If something fails, don't panic. Please go read the respective workflow's logs and
     configuration file to reverse-engineer your way to a fix/solution.

Congratulations! You've successfully cut a new release of _Black_. Go and stand up and
take a break, you deserve it.

```{important}
Once the release artifacts reach PyPI, you may see new issues being filed indicating
regressions. While regressions are not great, they don't automatically mean a hotfix
release is warranted. Unless the regressions are serious and impact many users, a hotfix
release is probably unnecessary.

In the end, use your best judgement and ask other maintainers for their thoughts.
```

> ### Changelog template

Author: In general I reordered the sections to what a regular user would find the most useful. Integrations got moved down because they aren't particularly relevant to most users. Packaging/parser changes on the other hand could probably either affect a lot of people and/or would be quite "newsworthy" (like supporting new syntax). 

Oh and docs really do not need the attention its fourth place gave it.

Reviewer: very nice 👍🏿

[Supported platforms are documented in the FAQ](labels/mypyc-support).

These matrix jobs use [cibuildwheel] which handles the complicated task of building C
extensions for many environments for us. Since building these wheels is slow, there are
multiple mypyc wheels jobs (hence the term "matrix") that build for a specific platform
(as noted in the job name in parentheses).

…

These jobs then use [twine] to upload the built distributions to PyPI.

#### Update stable branch

So this job doesn't _really_ belong here, but updating the `stable` branch after the
other PyPI jobs pass (they must pass for this job to start) makes the most sense. This
saves us from remembering to update the branch sometime after cutting the release.

- _Currently this workflow uses an API token associated with @ambv's PyPI account_

### Publish executables

This workflow builds native executables for multiple platforms using [PyInstaller]. This
allows people to download the executable for their platform and run _Black_ without a
[Python runtime](https://wiki.python.org/moin/PythonImplementations) installed.

The created binaries are stored on the associated GitHub Release for download over _IPv4

> only_ (GitHub still does not have IPv6 access 😢).

Reviewer 2: Still makes me sad every time I read / get bit by it :(

### docker

This workflow uses the QEMU powered `buildx` feature of Docker to upload an `arm64` and
`amd64`/`x86_64` build of the official _Black_ Docker image™.

> - _Currently this workflow uses an API Token associated with @cooperlees account_

Reviewer 2: Don't know if it's worth noting, but Ee from PSF also has access should I disappear.

Author: I'm pretty Ee is our contact should we lose access to any of the platforms we use haha. They control the psf GitHub organisation, and are an admin of PyPI and the pyfound dockerhub organisation. It's probably not worth noting IMO.

<!-- Include any especially major or disruptive changes here -->

> ### Stable style

Author: Thought it was worth making the distinction even clearer.

<!-- Changes that affect Black's stable style -->

### Preview style

<!-- Changes that affect Black's preview style -->

### Configuration

<!-- Changes to how Black can be configured -->

### Packaging

<!-- Changes to how Black is packaged, such as dependency requirements -->

### Parser

<!-- Changes to the parser or to version autodetection -->

### Performance

<!-- Changes that improve Black's performance. -->

### Output

<!-- Changes to Black's terminal output and error messages -->

### _Blackd_

<!-- Changes to blackd -->

### Integrations

<!-- For example, Docker, GitHub Actions, pre-commit, editors -->

### Documentation

<!-- Major changes to documentation and policies. Small docs changes
     don't need a changelog entry. -->
```

## Release workflows

> All of _Black_'s release automation use [GitHub Actions]. All workflows are therefore

Reviewer 3: ```suggestion
All of _Black_'s release automation uses [GitHub Actions]. All workflows are therefore
```

configured using YAML files in the `.github/workflows` directory of the _Black_
repository.

They are triggered by the publication of a [GitHub Release].

Below are descriptions of our release workflows.

### Publish to PyPI

> This is our main workhorse workflow. It builds a [sdist] and [wheels] to upload to PyPI

Reviewer 3: ```suggestion
This is our main workhorse workflow. It builds an [sdist] and [wheels] to upload to PyPI
```
I pronounced it "ess-dist", so it should have "an"

where the vast majority of users will download Black from. It's divided into three job
groups:

#### sdist + pure wheel

> This single job builds the sdist and pure Python wheel (ie. a wheel that only contains

Reviewer 3: ```suggestion
This single job builds the sdist and pure Python wheel (i.e., a wheel that only contains
```

Python code) using [build]. These artifacts are general-purpose and can be used on
basically any platform supported by Python.

> #### mypyc wheels (…)

Reviewer 3: What is the ellipsis for?

Author: Supposed to be a placeholder for the names the matrix jobs will have in practice. 
![image](https://user-images.githubusercontent.com/63936253/187310718-df6d606f-9b90-4363-925b-ed4afff334d0.png)

I'm happy to remove them if you find them confusing.

> We use [mypyc] to compile _Black_ into a CPython C Extension for significantly improved

Reviewer 3: ```suggestion
We use [mypyc] to compile _Black_ into a CPython C extension for significantly improved
```

performance. Wheels built with mypyc are platform and Python version specific.
[Supported platforms are documented in the FAQ](labels/mypyc-support).

These matrix jobs use [cibuildwheel] which handles the complicated task of building C
extensions for many environments for us. Since building these wheels is slow, there are
multiple mypyc wheels jobs (hence the term "matrix") that build for a specific platform
(noted in the job name in parentheses).

> ...

Reviewer 3: Same

Author: I was trying to add some separation between the following sentence and the previous section since the sdist / pure wheel job also uses twine to upload the distributions. 
![image](https://user-images.githubusercontent.com/63936253/187311079-ceb73efc-c45d-40d2-b494-4d1a5b9d4df6.png)

Happy to take suggestions for better ways of making this clearer.

Author: Resolved in https://github.com/psf/black/pull/3242/commits/82b3765cdc39a97569783c635d84ace758f928c6.

These jobs then use [twine] to upload the built distributions to PyPI.

#### Update stable branch

So this job doesn't _really_ belong here, but updating the `stable` branch after the
other PyPI jobs pass (they must pass for this job to start) makes the most sense. This
saves us from remembering to update the branch sometime after cutting the release.

- _Currently this workflow uses an API token associated with @ambv's PyPI account_

### Publish executables

This workflow builds native executable for multiple platforms using [PyInstaller]. This
allows people to download the executable for their platform and run _Black_ without a
[Python runtime](https://wiki.python.org/moin/PythonImplementations) installed.

> The created binaries are attached/stored on the associated GitHub Release for download

Reviewer 3: ```suggestion
The created binaries are stored on the associated GitHub Release for download
```

over _IPv4 only_ (GitHub still does not have IPv6 access 😢).

### docker

> This workflow uses the QEMU powered `buildx` feature of Docker to upload a `arm64` and

Reviewer 3: ```suggestion
This workflow uses the QEMU powered `buildx` feature of Docker to upload an `arm64` and
```

`amd64`/`x86_64` build of the official _Black_ Docker image™.

- _Currently this workflow uses an API Token associated with @cooperlees account_

```{note}
This also runs on each push to `main`.
```

[black-actions]: https://github.com/psf/black/actions
[build]: https://pypa-build.readthedocs.io/

> [calvar]: https://calver.org

Reviewer: ```suggestion
[calver]: https://calver.org
```
